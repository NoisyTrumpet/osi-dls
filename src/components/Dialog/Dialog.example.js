import React from 'react';
import PropTypes from 'prop-types';

import ExposedFeatureTrigger from './Fragments/ExposedFeatureTrigger';
import { Button } from '../Button';
import { Dialog, DialogManager } from '../Dialog';
import { IconButton } from '../IconButton';

/**
 *  Fully worked out example of how a Dialog is utilized by triggers anywhere in it's component tree
 */

// First, create a Context for every sub-app wide Dialog (ex. FilterFlyout)
const ExampleDialogContext = React.createContext(null);

const DialogExample = ({
  accentIcon,
  accentIconColor,
  hasGrayBackground,
  hideOverlay,
  type
}) => {
  // Use `<Dialog />` component with the necessary props, and set `isOpen` and `close` to the useContext values respectively
  const ExampleDialog = () => {
    const { isOpen, close } = React.useContext(ExampleDialogContext);
    return (
      <Dialog
        accentIcon={accentIcon}
        accentIconColor={accentIconColor}
        type={type}
        hasGrayBackground={hasGrayBackground}
        hideOverlay={hideOverlay}
        isOpen={isOpen}
        onClose={close}
        testId="modal"
      >
        <p>Tab through all the elements of the dialog to see how the focus traps works.</p>
        <a href="https://google.com">Link A</a>
      &nbsp;
        <a href="https://google.com">Link B</a>
        <br />
        <br />
        <Button mode="secondary">Button A</Button>
        <Button click={close}> Button B (extra close onClick)</Button>
      </Dialog>
    );
  };

  // Before rendering, call `DialogManager` to handle state and to wrap it in the ContextProvider
  const TriggerableDialog = DialogManager(ExampleDialog, ExampleDialogContext);

  // Create a dialog Trigger with useContext value of `open` as the onClick method
  const DialogTrigger = () => {
    const { open } = React.useContext(ExampleDialogContext);
    return (
      <IconButton click={open} icon="placeholder" variant="circular">Dialog Trigger</IconButton>
    );
  };

  const DialogTrigger2 = () => {
    const { open } = React.useContext(ExampleDialogContext);
    return (
      <Button click={open} mode="text-link">Dialog Trigger 2</Button>
    );
  };

  const DialogTrigger3 = () => {
    const { open } = React.useContext(ExampleDialogContext);
    return (
      <Button click={open} mode="quick-action-activated">Dialog Trigger 3</Button>
    );
  };

  const DialogTrigger4 = () => {
    const { open } = React.useContext(ExampleDialogContext);
    return (
      <Button click={open} mode="secondary">Dialog Trigger 4</Button>
    );
  };

  const DialogTrigger5 = () => {
    const { open } = React.useContext(ExampleDialogContext);
    return (
      <Button click={open} mode="primary">Dialog Trigger 5</Button>
    );
  };

  // Dialog is then contextually triggered by any child within it's component tree
  return (
    <TriggerableDialog>
      <div style={{ paddingBottom: '8px' }}>
        <DialogTrigger />
      </div>
      <ul>
        <li style={{ paddingBottom: '8px' }}><DialogTrigger2 /></li>
        <li style={{ paddingBottom: '8px', textAlign: 'center' }}>some text separating unique trigger elements…</li>
        <li style={{ paddingBottom: '8px' }}><DialogTrigger3 /></li>
        <li style={{ paddingBottom: '8px', textAlign: 'center' }}>some text separating unique trigger elements…</li>
        <li style={{ paddingBottom: '8px' }}><DialogTrigger4 /></li>
      </ul>
      <div style={{ paddingBottom: '8px' }}>
        <DialogTrigger5 />
      </div>
    </TriggerableDialog >
  );
};

const ExposedTriggerExample = ({
  accentIcon,
  accentIconColor,
  hasGrayBackground,
  hideOverlay,
  type
}) => {
  // Use `<Dialog />` component with the necessary props, and set `isOpen` and `close` to the useContext values respectively
  const ExampleDialog = () => {
    const { isOpen, close } = React.useContext(ExampleDialogContext);
    return (
      <Dialog
        type={type}
        accentIcon={accentIcon}
        exposedTrigger
        hasGrayBackground={hasGrayBackground}
        hideOverlay={hideOverlay}
        isOpen={isOpen}
        onClose={close}
        testId="modal"
      >
        <p>Tab through all the elements of the dialog to see how the focus traps works.</p>
        <a href="https://google.com">Link A</a>
      &nbsp;
        <a href="https://google.com">Link B</a>
        <br />
        <br />
        <Button mode="secondary">Button A</Button>
        <Button click={close}> Button B (extra close onClick)</Button>
      </Dialog>
    );
  };

  // Before rendering, call `DialogManager` to handle state and to wrap it in the ContextProvider
  const TriggerableDialog = DialogManager(ExampleDialog, ExampleDialogContext);

  // Dialog is then contextually triggered by any child within it's component tree
  return (
    <TriggerableDialog>
      <ExposedFeatureTrigger
        accentIcon={accentIcon}
        accentIconColor={accentIconColor}
        context={ExampleDialogContext}
        exposedText='Dialog Window'
        hasGrayBackground={hasGrayBackground}
        hideOverlay={hideOverlay}
        testId="featureTrigger"
      />
    </TriggerableDialog >
  );
};

DialogExample.propTypes = {
  accentIcon: PropTypes.string,
  accentIconColor: PropTypes.string,
  hasGrayBackground: PropTypes.bool,
  hideOverlay: PropTypes.bool,
  type: PropTypes.string,
};

ExposedTriggerExample.propTypes = {
  accentIcon: PropTypes.string,
  accentIconColor: PropTypes.string,
  hasGrayBackground: PropTypes.bool,
  hideOverlay: PropTypes.bool,
  type: PropTypes.string,
};

export { DialogExample, ExposedTriggerExample };
