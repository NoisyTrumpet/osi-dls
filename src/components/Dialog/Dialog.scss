@import './DialogAnimations.scss';
@import './DialogExposedTrigger.scss';

.dialog-overlay {
  display: none;
  position: fixed;
  background: rgba(map-get($colors, 'brand-black'), 0.64);
  align-items: center;
  justify-content: center;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
}

.hide-overlay .dialog-overlay {
  background: rgba(map-get($colors, 'brand-black'), 0);
}

[data-open='true'] {
  .dialog-overlay {
    display: flex;

    .dialog-wrapper {
      display: flex;
      flex-direction: column;
    }
  }

  .dialog-scroll {
    overflow-y: auto;
  }
}

/**
 * Dialog - Modal
 */
[data-type='modal'] {
  background: map-get($colors, 'brand-white');
  border-radius: $space-4;
  box-shadow: $shadow-2;
  width: 100%;
  max-height: calc(100vh - 102px - 160px);
  max-width: rem(448px);
  opacity: 0;
  transform: translate(0, $space-32);

  @include media($mq-mobile) {
    margin: $space-56 $space-24;
  }

  .dialog-content {
    padding: $space-32;
  }

  .dialog-title-icon {
    display: inline-block;
    margin: $space-12;
  }

  .dialog-close {
    position: absolute;
    top: -$space-12;
    right: $space-16;

    &:after {
      content: '';
      position: absolute;
      top: 0;
      right: 0;
      bottom: 0;
      left: 0;
      z-index: 1;
    }
  }

  .typography-headline {
    display: inline-block;
    margin-bottom: $space-16;
  }
}

/**
 * Dialog - Feature Modal
 */
[data-type='feature-modal'] {
  background: map-get($colors, 'brand-white');
  position: absolute;
  bottom: 0;
  top: auto;
  overflow-y: visible;
  max-height: calc(100vh - 48px - 56px);
  max-width: none;
  width: 100%;
  text-align: center;

  &.dialog-wrapper-gray {
    background: map-get($colors, 'gray0');

    .dialog-top-curve g,
    .dialog-bottom-curve g {
      fill: map-get($colors, 'gray0');
    }
  }

  .dialog-top-curve g,
  .dialog-bottom-curve g {
    fill: map-get($colors, 'brand-white');
  }

  .dialog-title {
    width: 100%;
    text-align: center;
    margin-bottom: $space-16;
    margin-top: $space-8;
  }

  .dialog-content {
    padding: 0 $space-24 rem(10px) $space-24;
  }

  .dialog-accent-icon {
    position: absolute;
    top: -6.4vw;
    left: 50%;
    transform: translate(-50%, -50%);
  }

  .dialog-close {
    position: absolute;
    right: rem(20px);
    top: rem(-20px);

    &:after {
      content: '';
      position: absolute;
      top: 0;
      right: 0;
      bottom: 0;
      left: 0;
      z-index: 1;
    }
  }

  .dialog-top-curve {
    transform: translate(0, -100%);
    margin-top: 1px; // Fixes the space between the dialog and the svg
    display: block;
    position: absolute;
    width: 100vw;
    height: 6.4vw;
  }

  @include media($mq-tablet-up) {
    bottom: auto;
    max-height: calc(100vh - 102px - 160px);
    max-width: rem(475px);

    .dialog-accent-icon {
      position: absolute;
      top: - $space-48;
      left: 50%;
      transform: translate(-50%, -50%);
    }

    .dialog-close {
      top: rem(-42px);
    }

    .dialog-title {
      display: inline-block;
    }

    .dialog-top-curve {
      width: 100%;
      max-width: rem(475px);
      height: auto;
    }

    .dialog-bottom-curve {
      position: absolute;
      bottom: 2px;
      display: block;
      transform: translate(0, 100%);
      width: 100%;
      max-width: rem(475px);
    }
  }
}

[data-type='feature-modal'] {
  &:not(.exposed-trigger) {
    opacity: 0;
    transform: translate(0, calc(100% - 52px));
    @include media($mq-tablet-up) {
      transform: translate(0, $space-32);
    }
  }
}

/**
 * Dialog - Toast
 */
[data-type='toast'] {
  background: map-get($colors, 'brand-white');
  border-radius: $space-8;
  box-shadow: $shadow-1;
  padding: $space-16;
  margin: $space-24;
  width: 100%;
  max-width: rem(672px);
  opacity: 0;
  transform: translate(0, $space-32);

  @include media($mq-mobile) {
    margin: $space-16;
  }

  .dialog-close {
    position: absolute;
    top: $space-16;
    right: $space-16;

    &:after {
      content: '';
      position: absolute;
      top: 0;
      right: 0;
      bottom: 0;
      left: 0;
      z-index: 1;
    }
  }

  .typography-headline {
    margin-bottom: $space-4;
  }
}

/**
 * Dialog - Drawer (left/right/bottom)
 */
[data-type^='drawer'] {
  @media only screen and (max-width: 407px) {
    max-width: calc(100% - 32px);
  }

  @media only screen and (max-width: 375px) {
    max-width: calc(100% - 24px);
  }

  background: map-get($colors, 'gray0');
  box-shadow: $shadow-1;
  padding: 0;
  width: 100%;
  max-width: rem(375px);
  opacity: 0;
  position: absolute;
  top: 0;
  bottom: 0;

  &[data-type$='-left'] {
    left: 0;
    transform: translate(-100%, 0);

    .dialog-title {
      @extend %screen-reader-text;
    }
  }

  &[data-type$='-right'] {
    right: 0;
    transform: translate(100%, 0);

    .dialog-title {
      @extend %screen-reader-text;
    }
  }

  &[data-type$='-bottom'] {
    overflow-y: visible;
    transform: translate(0, 100%);
    background: map-get($colors, 'brand-white');
    margin-top: $space-48;
    max-width: none;
    max-height: calc(100vh - 48px - 56px);
    width: 100%;
    top: auto;
    text-align: center;

    .dialog-top-curve g {
      fill: map-get($colors, 'brand-white');
    }

    .dialog-title {
      margin-top: $space-12;
    }

    .dialog-title-icon {
      position: absolute;
      top: -$space-24;
      left: 50%;
      transform: translate(-50%, -50%);
    }

    .dialog-close {
      position: absolute;
      top: rem(-20px);
      right: rem(20px);

      &:after {
        content: '';
        position: absolute;
        top: 0;
        right: 0;
        bottom: 0;
        left: 0;
        z-index: 1;
      }
    }

    .dialog-top-curve {
      transform: translate(0, -100%);
      margin-top: 1px; // Fixes the space between the dialog and the svg
      display: block;
      position: absolute;
    }
  }

  &[data-type$='-top'] {
    overflow-y: visible;
    transform: translate(0, -100%);
    background: map-get($colors, 'brand-white');
    margin-bottom: $space-48;
    max-width: none;
    max-height: calc(100vh - 48px - 56px);
    width: 100%;
    bottom: auto;
    text-align: center;

    .dialog-top-curve g {
      fill: map-get($colors, 'brand-white');
    }

    .dialog-top-curve {
      transform: translate(0, -100%) rotate(180deg);
      bottom: -$space-48;
      margin-top: 1px; // Fixes the space between the dialog and the svg
      display: block;
      position: absolute;

    }

    .dialog-title {
      margin-top: $space-12;
    }

    .dialog-title-icon {
      position: absolute;
      bottom: rem(-70px);
      left: 50%;
      transform: translate(-50%, -50%);
    }

    .dialog-close {
      position: absolute;
      top: $space-12;
      right: $space-24;
      border: none;

      &:hover, &:focus {
        box-shadow: none;
      }

      &:after {
        content: '';
        position: absolute;
        top: 0;
        right: 0;
        bottom: 0;
        left: 0;
        z-index: 1;
      }
    }

    .dialog-bottom-curve {
      transform: translate(0, 100%);
      margin-top: 1px; // Fixes the space between the dialog and the svg
      display: block;
      position: absolute;
    }
  }
}

// drawer-left / drawer-right top spacing
[data-animate='drawer-right'],
[data-animate='drawer-left'] {
  .dialog-overlay {
    top: $space-56;
  }
}
