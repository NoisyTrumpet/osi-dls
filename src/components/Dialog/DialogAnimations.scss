[data-open='true'] {
  // Modal and Toast Animations
  &[data-animate='modal'] .dialog-wrapper,
  &[data-animate='feature-modal'] .dialog-wrapper:not(.exposed-trigger),
  &[data-animate='toast'] .dialog-wrapper {
    animation: modalSlide 0.16s linear forwards 0.08s;
  }

  // Drawer Animations
  &[data-animate='drawer-left'] .dialog-wrapper {
    animation: openFromLeft 0.24s linear forwards 0.08s;
  }

  &[data-animate='drawer-right'] .dialog-wrapper {
    animation: openFromRight 0.24s linear forwards 0.08s;
  }

  &[data-animate='drawer-bottom'] .dialog-wrapper {
    animation: openFromBottom 0.24s linear forwards 0.08s;
  }

  &[data-animate='drawer-top'] .dialog-wrapper {
    animation: openFromTop 0.24s linear forwards 0.08s;
  }

  &[data-animate='feature-modal'] {
    .dialog-wrapper.exposed-trigger {
      animation: exposedTriggerOpenMobile 0.24s linear forwards 0.08s;
      
      @include media($mq-tablet-up) {
        animation: exposedTriggerOpen 0.24s linear forwards 0.08s;
      }
    }
  }

  @include media($mq-mobile) {
    &[data-animate='feature-modal'] .dialog-wrapper:not(.exposed-trigger) {
      animation: openFromBottom 0.24s linear forwards 0.08s;
    }
  }
}

@keyframes exposedTriggerOpen {
  from {
    background: map-get($colors, 'brand-white');
    bottom: 0;
  }

  to {
    background: map-get($colors, 'gray0');
    bottom: 50%;
    transform: translate(0, 50%);
  }
}

@keyframes exposedTriggerOpenMobile {
  from {
    background: map-get($colors, 'brand-white');
    transform: translate(0, calc(100% - 52px));
  }

  to {
    background: map-get($colors, 'gray0');
    transform: translate(0, 0);
  }
}

@keyframes openFromBottom {
  from {
    opacity: 0;
    transform: translate(0, 100%);
  }

  to {
    opacity: 1;
    transform: translate(0, 0);
  }
}

@keyframes openFromTop {
  from {
    opacity: 0;
    transform: translate(0, -100%);
  }

  to {
    opacity: 1;
    transform: translate(0, 0);
  }
}

@keyframes openFromLeft {
  from {
    opacity: 0;
    transform: translate(-100%, 0);
  }

  to {
    opacity: 1;
    transform: translate(0, 0);
  }
}

@keyframes openFromRight {
  from {
    opacity: 0;
    transform: translate(100%, 0);
  }

  to {
    opacity: 1;
    transform: translate(0, 0);
  }
}

@keyframes modalSlide {
  from {
    opacity: 0;
    transform: translate(0, $space-32);
  }

  to {
    opacity: 1;
    transform: translate(0, 0);
  }
}