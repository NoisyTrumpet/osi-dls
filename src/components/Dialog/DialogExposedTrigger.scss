/**
 * Dialog - Feature Modal Exposed Trigger
 */
.feature-modal-exposed-trigger {
  background: map-get($colors, 'brand-white');
  position: fixed;
  bottom: 0;
  left: 0;
  right: 0;
  top: auto;
  overflow-y: visible;
  transform: translate(0, calc(100% - 52px));
  max-height: calc(100vh - 48px - 56px);
  max-width: none;
  width: 100%;
  text-align: center;
  filter: drop-shadow(0 0.25rem 0.5rem rgba(0, 0, 0, 0.16));
  padding-bottom: $space-16;
  opacity: 1;

  &:hover,
  &.focused {
    filter: drop-shadow(0 0.25rem 0.5rem rgba(0, 0, 0, 0.24));
    cursor: pointer;
  }

  .dialog-top-curve g,
  .dialog-bottom-curve g {
    fill: map-get($colors, 'brand-white');
  }

  &.dialog-wrapper-gray {
    background: map-get($colors, 'gray0');

    .dialog-top-curve g,
    .dialog-bottom-curve g {
      fill: map-get($colors, 'gray0');
    }
  }

  .dialog-title {
    width: 100%;
    text-align: center;
    margin-bottom: $space-16;
    margin-top: $space-8;
    padding-left: $space-24;
    padding-right: $space-24;
  }

  .dialog-accent-icon {
    position: absolute;
    top: -6.4vw;
    left: 50%;
    transform: translate(-50%, -50%);
  }

  .dialog-close {
    position: absolute;
    right: rem(20px);
    top: rem(-20px);

    &:after {
      content: '';
      position: absolute;
      top: 0;
      right: 0;
      bottom: 0;
      left: 0;
      z-index: 1;
    }
  }

  .dialog-top-curve {
    transform: translate(0, -100%);
    margin-top: 1px; // Fixes the space between the dialog and the svg
    display: block;
    position: absolute;
    width: 100vw;
    height: 6.4vw;
    top: 0;
  }

  @include media($mq-tablet-up) {
    margin: 0 auto;
    max-height: calc(100vh - 102px - 160px);
    max-width: rem(475px);

    .dialog-accent-icon {
      position: absolute;
      top: - $space-48;
      left: 50%;
      transform: translate(-50%, -50%);
    }

    .dialog-close {
      top: rem(-42px);
    }

    .dialog-title {
      display: inline-block;
    }

    .dialog-top-curve {
      width: 100%;
      max-width: rem(475px);
      height: auto;
    }

    .dialog-bottom-curve {
      position: absolute;
      bottom: 0;
      display: block;
      transform: translate(0, 100%);
      width: 100%;
      max-width: rem(475px);
    }
  }
}

[data-type='feature-modal'].exposed-trigger {
  opacity: 1;
  transform: translate(0, calc(100% - 52px));

  @include media($mq-tablet-up) {
    bottom: 0;
  }
}