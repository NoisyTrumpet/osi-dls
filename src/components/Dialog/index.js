export { default as Dialog } from './Dialog'; // eslint-disable-line import/prefer-default-export
export { default as DialogManager } from './HoC/DialogManager';
