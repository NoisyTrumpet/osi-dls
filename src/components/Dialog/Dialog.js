import React, { useEffect, useRef } from "react";
import ReactDOM from "react-dom";
import classNames from "classnames";
import PropTypes from "prop-types";
import { useMediaQuery } from "react-responsive";
import { trapFocus, canUseDOM, bodyScroll } from "../../util";
import { AccentIcon } from "../AccentIcon";
import { IconButton } from "../IconButton";
import { InlineSVG } from "../InlineSVG";
import { Typography } from "../Typography";
import "./Dialog.scss";

import {
  AccentCircle,
  MobileModalCurveTop,
  DesktopModalCurveBottom,
  DesktopModalCurveTop
} from "../../assets/svg";

const Dialog = ({
  accentIcon,
  accentIconColor,
  addDialogClass,
  appendEl, // append an element to trapFocus in utils.js
  children,
  disableTrapFocus, // disable trapFocus in utils.js
  exposedTrigger,
  hasGrayBackground,
  hideOverlay,
  id,
  isOpen,
  onClose,
  prependEl, // prepend an element to trapFocus in utils.js
  testId,
  title,
  titleDisabled,
  type
}) => {
  // Unicode value of the `TAB` and `ESC` key
  const KEYCODE_ESC = 27;

  const dialog = useRef(null);
  const overlay = useRef(null);
  const close = useRef(null);

  let removeKeyDown;

  // classnames for dialog container
  const dialogClasses = classNames(
    "dialog-container",
    hideOverlay && "hide-overlay",
    addDialogClass
  );

  // classnames for dialog wrapper
  const dialogWrapperClasses = classNames(
    "dialog-wrapper",
    hasGrayBackground && "dialog-wrapper-gray",
    exposedTrigger && "exposed-trigger"
  );

  const triggerOnCloseCallback = () => {
    if (onClose) {
      onClose();
    }
  };

  // closeMethod (str) = how the close is being triggered
  // element = what user just clicked on to close the dialog and call this function
  const closeDialog = (closeMethod, element) => {
    if (closeMethod === "escape-key") {
      triggerOnCloseCallback();
    }
    if (closeMethod === "close-button" || closeMethod === "overlay") {
      const exclude = element.target.classList;
      if (
        exclude.contains("dialog-overlay") ||
        exclude.contains("dialog-close")
      ) {
        triggerOnCloseCallback();
      }
    }
    if (removeKeyDown) {
      removeKeyDown();
      removeKeyDown = null;
    }
  };

  // Update page when the 'isOpen' prop changes.
  // 'isOpen' is controlled in the parent component,
  // so true here means "please open the dialog."
  useEffect(() => {
    if (isOpen && !disableTrapFocus) {
      // need a little time for the dialog to appear to safely focus
      setTimeout(() => {
        removeKeyDown = trapFocus(dialog.current, prependEl, appendEl);
        bodyScroll(false);
      }, 100);
    } else {
      closeDialog("initial-trigger");
      bodyScroll(true);
    }
  }, [isOpen]);

  const onKeyDown = e => e.keyCode === KEYCODE_ESC && closeDialog("escape-key");

  const isMobile = useMediaQuery({ query: "(max-width: 767px)" });
  const DialogComponent = () => (
    <div
      className={dialogClasses}
      id={id}
      data-testid={testId}
      data-open={isOpen}
      data-animate={type}
      aria-hidden={!isOpen}
      ref={dialog}
    >
      <div
        className="dialog-overlay"
        tabIndex="-1"
        onClick={e => closeDialog("overlay", e)}
        role="presentation"
        ref={overlay}
      >
        <div // eslint-disable-line jsx-a11y/no-noninteractive-element-interactions
          className={dialogWrapperClasses}
          data-type={type}
          aria-modal="true"
          aria-labelledby={`${id}-title`}
          role="dialog"
          onKeyDown={e => onKeyDown(e)}
        >
          <div className="dialog-scroll">
            {(type === "drawer-bottom" || type === "feature-modal") && (
              <InlineSVG
                path={
                  type === "drawer-bottom" || isMobile
                    ? `${MobileModalCurveTop}`
                    : `${DesktopModalCurveTop}`
                }
                title="Dialog Top Curve"
                description={isMobile ? "Drawer Curve" : "Top of Modal Curve"}
                role="img"
                focus={false}
                className="dialog-top-curve"
              />
            )}

            {type === "feature-modal" && (
              <InlineSVG
                path={
                  isMobile
                    ? `${MobileModalCurveTop}`
                    : `${DesktopModalCurveTop}`
                }
                title="Dialog Bottom Curve"
                description={
                  isMobile ? "Drawer Curve" : "Bottom of Modal Curve"
                }
                role="img"
                focus={false}
                className="dialog-top-curve"
              />
            )}

            {(type === "feature-modal" || type === "drawer-bottom") &&
              accentIcon && (
                <div
                  className={
                    type === "feature-modal"
                      ? "dialog-accent-icon"
                      : "dialog-title-icon"
                  }
                >
                  <AccentIcon
                    accent={AccentCircle}
                    icon={accentIcon}
                    iconColor={accentIconColor}
                  />
                </div>
              )}

            {!titleDisabled && (
              <Typography
                id={`${id}-title`}
                variant={type === "toast" ? "subtext" : "headline-5"}
                addClass="dialog-title"
              >
                {title}
              </Typography>
            )}

            <div id={`${id}-content`} className="dialog-content">
              {children}
            </div>

            {type === "feature-modal" && !isMobile && (
              <InlineSVG
                path={`${DesktopModalCurveBottom}`}
                title="Dialog Bottom Curve"
                description="Top of Modal Curve"
                role="img"
                focus={false}
                className="dialog-bottom-curve"
              />
            )}
          </div>

          {type !== "drawer-left" && type !== "drawer-right" && (
            <IconButton
              click={e => closeDialog("close-button", e)}
              icon="menuclose-bold"
              addClass="dialog-close"
              variantTypo="srt"
              ref={close}
              variant={type === "toast" ? "simple" : "tiny"}
              testId="closeDialogButton"
            >
              {`Close ${title}`}
            </IconButton>
          )}
        </div>
      </div>
    </div>
  );

  if (canUseDOM()) {
    // use createPortal to place the dialog at the end of the body tag
    return ReactDOM.createPortal(<DialogComponent />, document.body);
  }
  return <DialogComponent />;
};

Dialog.defaultProps = {
  id: `dialog-${Math.floor(Math.random() * 10000)}`,
  type: "modal",
  title: "Dialog Window"
};

Dialog.propTypes = {
  accentIcon: PropTypes.string,
  accentIconColor: PropTypes.string,
  addDialogClass: PropTypes.string,
  appendEl: PropTypes.string, // appendEl="button.dialog-close"
  children: PropTypes.node,
  disableTrapFocus: PropTypes.bool,
  exposedTrigger: PropTypes.bool,
  hasGrayBackground: PropTypes.bool,
  hideOverlay: PropTypes.bool,
  id: PropTypes.string,
  isOpen: PropTypes.bool,
  onClose: PropTypes.func,
  prependEl: PropTypes.string,
  testId: PropTypes.string,
  type: PropTypes.oneOf([
    "modal",
    "feature-modal",
    "toast",
    "drawer-left",
    "drawer-right",
    "drawer-bottom",
    "drawer-top"
  ]),
  title: PropTypes.string,
  titleDisabled: PropTypes.bool
};

export default Dialog;
