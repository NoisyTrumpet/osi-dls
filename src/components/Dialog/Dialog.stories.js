import React from "react";
import { storiesOf } from "@storybook/react";
import { withSmartKnobs } from "storybook-addon-smart-knobs";
import { select, boolean } from "@storybook/addon-knobs";
import iconArray from "../Icon/helpers/getIconArray";
import { Dialog } from "../Dialog";
import { DialogExample, ExposedTriggerExample } from "./Dialog.example";

const parameters = {
  component: Dialog,
  componentSubtitle: "MicroModal powered dialogs",
  jest: ["Dialog"]
};

storiesOf("components | Dialog", module)
  .addParameters(parameters)
  .addDecorator(withSmartKnobs({ ignoreProps: ["type", "children"] }))
  .add("Modal", () => (
    <DialogExample
      accentIcon={select("icon", iconArray, "achievement-bold")}
      hideOverlay={boolean("hideOverlay", false)}
      type="modal"
    />
  ))
  .add("Feature Modal", () => (
    <DialogExample
      accentIcon={select("icon", iconArray, "achievement-bold")}
      hideOverlay={boolean("hideOverlay", false)}
      type="feature-modal"
    />
  ))
  .add("Feature Modal - Exposed Trigger", () => (
    <ExposedTriggerExample
      accentIcon={select("icon", iconArray, "achievement-bold")}
      hasGrayBackground={boolean("hasGrayBackground", true)}
      hideOverlay={boolean("hideOverlay", false)}
      type="feature-modal"
    />
  ))
  .add("Drawer - Left", () => <DialogExample type="drawer-left" />)
  .add("Drawer - Right", () => <DialogExample type="drawer-right" />)
  .add("Drawer - Bottom", () => (
    <DialogExample
      accentIcon={select("icon", iconArray, "achievement-bold")}
      hideOverlay={boolean("hideOverlay", false)}
      type="drawer-bottom"
    />
  ))
  .add("Drawer - Top", () => (
    <DialogExample
      accentIcon={select("icon", iconArray, "achievement-bold")}
      hideOverlay={boolean("hideOverlay", false)}
      type="drawer-top"
    />
  ));
