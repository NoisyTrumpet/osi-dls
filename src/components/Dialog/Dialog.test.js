import React from 'react';
import {
  render,
  fireEvent, 
  cleanup,
} from '@testing-library/react';
import {
  toBeInTheDocument,
  toHaveClass,
  toHaveAttribute,
  toBeVisible,
} from '@testing-library/jest-dom/matchers';

import { DialogExample, ExposedTriggerExample } from './Dialog.example'

expect.extend({
  toBeInTheDocument,
  toHaveClass,
  toHaveAttribute,
  toBeVisible,
});

describe('<DialogExample />', () => {
  test('renders Dialog component', () => {
    const { container, getAllByText, getByText } = render(
      <DialogExample
        accentIcon={{ name: 'achievement-bold' }}
        type="modal"
      />
    );

    expect(getAllByText(/Dialog Trigger/i)).toHaveLength(5);
    expect(getAllByText('Dialog Window')).toHaveLength(1);
    expect(getByText('Close Dialog Window')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  afterEach(cleanup);
});

describe('<FeatureModalExample />', () => {
  test('renders Feature Modal component', () => {
    const { container, getAllByText, getByText } = render(
      <ExposedTriggerExample
        type="feature-modal"
        accentIcon={{ name: 'achievement-bold' }}
      />
    );

    expect(getAllByText('Dialog Window')).toHaveLength(2);
    expect(getByText('Close Dialog Window')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  test('opens and closes Dialog Window', () => {
    const { getByTestId } = render(
      <ExposedTriggerExample
        type="feature-modal"
        accentIcon={{ name: 'achievement-bold' }}
      />
    );

    const featureTrigger = getByTestId('featureTrigger');
    expect(featureTrigger).toBeInTheDocument();

    const modal = getByTestId('modal');
    expect(modal).toBeInTheDocument();
    expect(modal).toHaveAttribute('data-open', 'false');

    fireEvent.click(featureTrigger);
    expect(getByTestId('modal')).toHaveAttribute('data-open', 'true');

    const closeModal = getByTestId('modal').getElementsByClassName('dialog-close');
    expect(closeModal[0]).toBeInTheDocument();

    fireEvent.click(closeModal[0]);
    expect(getByTestId('modal')).toHaveAttribute('data-open', 'false');
  });

  test('hides Feature Modal Exposed Trigger', () => {
    const { getByTestId } = render(
      <ExposedTriggerExample
        type="feature-modal"
        accentIcon={{ name: 'achievement-bold' }}
      />
    );

    const featureTrigger = getByTestId('featureTrigger');
    expect(featureTrigger).toBeInTheDocument();

    fireEvent.click(getByTestId('closeExposedTrigger'));
    expect(featureTrigger).not.toBeInTheDocument();
  });

  afterEach(cleanup);
});
