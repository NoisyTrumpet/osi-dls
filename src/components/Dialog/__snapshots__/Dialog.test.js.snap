// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DialogExample /> renders Dialog component 1`] = `
<div>
  <div
    style="padding-bottom: 8px;"
  >
    <button
      class="btn-icon btn-icon-circular"
      type="button"
    >
      <span
        class="screen-reader-text"
      >
        Dialog Trigger
      </span>
      <span
        aria-hidden="true"
        class="icon icon-placeholder icon-16"
        data-testid="icon"
      />
    </button>
  </div>
  <ul>
    <li
      style="padding-bottom: 8px;"
    >
      <button
        class="btn btn-text-link"
        type="button"
      >
        
        <span
          class="typography typography-text-link color-accentB"
        >
          Dialog Trigger 2
        </span>
        
      </button>
    </li>
    <li
      style="padding-bottom: 8px; text-align: center;"
    >
      some text separating unique trigger elements…
    </li>
    <li
      style="padding-bottom: 8px;"
    >
      <button
        class="btn btn-quick-action-activated"
        type="button"
      >
        
        <span
          class="typography typography-cta-tertiary color-primary"
        >
          Dialog Trigger 3
        </span>
        
      </button>
    </li>
    <li
      style="padding-bottom: 8px; text-align: center;"
    >
      some text separating unique trigger elements…
    </li>
    <li
      style="padding-bottom: 8px;"
    >
      <button
        class="btn btn-secondary"
        type="button"
      >
        
        <span
          class="typography typography-cta color-primary"
        >
          Dialog Trigger 4
        </span>
        
      </button>
    </li>
  </ul>
  <div
    style="padding-bottom: 8px;"
  >
    <button
      class="btn btn-primary"
      type="button"
    >
      
      <span
        class="typography typography-cta color-brand-white"
      >
        Dialog Trigger 5
      </span>
      
    </button>
  </div>
</div>
`;

exports[`<FeatureModalExample /> renders Feature Modal component 1`] = `
<div>
  <div
    class="feature-modal-exposed-trigger"
  >
    <div
      class="dialog-scroll"
      data-testid="featureTrigger"
      role="button"
      tabindex="0"
    >
      <div
        class="dialog-accent-icon"
      >
        <div
          class="accent-icon"
        >
          <div
            class="icon-wrapper"
          >
            <span
              aria-hidden="true"
              class="icon icon-[object Object] icon-brand-white icon-24"
              data-testid="triggerAccentIcon"
            />
          </div>
        </div>
      </div>
      <h6
        class="typography typography-headline typography-headline-6 dialog-title"
      >
        Dialog Window
      </h6>
    </div>
    <button
      class="btn-icon btn-icon-tiny dialog-close"
      data-testid="closeExposedTrigger"
      type="button"
    >
      <span
        class="screen-reader-text"
      >
        Close
      </span>
      <span
        aria-hidden="true"
        class="icon icon-menuclose-bold icon-16"
        data-testid="icon"
      />
    </button>
  </div>
</div>
`;
